src/bsoncxx/test/bson_get_values.cpp

```diff
@@ -0,0 +1,154 @@
+    using namespace builder::stream;
+
+    builder::stream::document build_doc;
+    // {
+    //     "ints": [ 1, 3, [ 5 ] ],
+    //     "bools": [ true, false ]
+    // }
+    build_doc << "ints" << open_array << 1 << 3 << open_array << 5 << close_array << close_array
+              << "bools" << open_array << true << false << close_array;
+
+    auto doc = build_doc.view();
+
+    SECTION("works with one level") {
+        REQUIRE(doc["ints"][0]);
+        REQUIRE(doc["ints"][0].get_int32() == 1);
+
+        REQUIRE(doc["ints"][1]);
+        REQUIRE(doc["ints"][1].get_int32() == 3);
+
+        REQUIRE(doc["bools"][0]);
+        REQUIRE(doc["bools"][0].get_bool() == true);
+
+        REQUIRE(doc["bools"][1]);
+        REQUIRE(doc["bools"][1].get_bool() == false);
+    }
+
+    SECTION("works with two levels") {
+        REQUIRE(doc["ints"][2][0]);
+        REQUIRE(doc["ints"][2][0].get_int32() == 5);
+    }
+
+    SECTION("throws on non-document") {
+        REQUIRE_THROWS(doc["ints"][0][0]);
+        REQUIRE_THROWS(doc["ints"][2][0][0]);
+        REQUIRE_THROWS(doc["bools"][1][0]);
+    }
+
+    SECTION("returns invalid on not found") {
+        REQUIRE(!doc["ints"][9]);
+        REQUIRE(!doc["ints"][2][9]);
+        REQUIRE(!doc["bools"][9]);
+    }
+}
+
+TEST_CASE("[] can reach into nested documents", "[bsoncxx]") {
+    using namespace builder::stream;
+
+    builder::stream::document build_doc;
+    // {
+    //     "ints": {
+    //         "x": 1,
+    //         "y": 3,
+    //         "more": {
+    //             "z": 5
+    //         }
+    //     },
+    //     "bools": {
+    //         "t": true,
+    //         "f": false,
+    //     }
+    // }
+    build_doc << "ints" << open_document << "x" << 1 << "y" << 3 << "more" << open_document << "z"
+              << 5 << close_document << close_document << "bools" << open_document << "t" << true
+              << "f" << false << close_document;
+
+    auto doc = build_doc.view();
+
+    SECTION("works with one level") {
+        REQUIRE(doc["ints"]["x"]);
+        REQUIRE(doc["ints"]["x"].get_int32() == 1);
+
+        REQUIRE(doc["ints"]["y"]);
+        REQUIRE(doc["ints"]["y"].get_int32() == 3);
+
+        REQUIRE(doc["bools"]["t"]);
+        REQUIRE(doc["bools"]["t"].get_bool() == true);
+
+        REQUIRE(doc["bools"]["f"]);
+        REQUIRE(doc["bools"]["f"].get_bool() == false);
+    }
+
+    SECTION("works with two levels") {
+        REQUIRE(doc["ints"]["more"]["z"]);
+        REQUIRE(doc["ints"]["more"]["z"].get_int32() == 5);
+    }
+
+    SECTION("throws on non-document") {
+        REQUIRE_THROWS(doc["ints"]["x"]["key"]);
+        REQUIRE_THROWS(doc["ints"]["more"]["z"]["key"]);
+        REQUIRE_THROWS(doc["bools"]["f"]["key"]);
+    }
+
+    SECTION("returns invalid on not found") {
+        REQUIRE(!doc["ints"]["badKey"]);
+        REQUIRE(!doc["ints"]["more"]["badKey"]);
+        REQUIRE(!doc["bools"]["badKey"]);
```

Comment:
Can you add a case for trying to reach into an invalid element?  Like `doc["ints"]["blah"]["blah"]`